File: c++/p004.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>
#include <string>
using namespace std;
bool check_palindromic(int n){
    string s = to_string(n);
    string s_r = string(s.rbegin(),s.rend());
    if (s==s_r){
        return true;    
    }
    else{
        return false;
    }
}
int palindromic_number(){
    int nmax = 0;
    for (int i = 999; i >=100; i--){
        if (i*i <nmax){
            break;
        }
        for (int j = i; j>=100 ; j--){
            int c = i * j;
            bool is_palindromic = check_palindromic(c);
            if (is_palindromic == true){
                if (c>nmax){
                    nmax = c;
                }
                break;
            }
        }
    }
    return nmax;
}
int main(int argc, char** argv){

    cout << "Project Euler 4" << endl;
    cout << palindromic_number()<<endl;
    return 0;
}

File: python/p004.py
import numpy as np
import sys
"""Find the largest palindrome made from the product of two 3-digit numbers."""
def palindromic_number_v1():
    ans = max(i*j
            for i in range(100,1000)
            for j in range(100,1000)
            if str(i*j) == str(i*j)[::-1])
    return ans

def palindromic_number_v2():
    nmax = 0
    for i in range(999,100,-1):    # dow from 999 to 100
        if i*i < nmax:             # check loop is less the sqrt of highest number  
            break
        for j in range(i,100,-1):  # further narrow search space
            n = i*j
            if check_palindromic(n):
                if n > nmax:
                    nmax = n
                break
    return nmax

def check_palindromic(n):
    n = str(n)
    if n == n[::-1]:
        return True
    else:
        return False


if __name__ == "__main__":
    print ('Project Euler 4')
    print (palindromic_number())
    n = palindromic_number()
    print (check_palindromic(n))
    print (palindromic_number_v2())
